trin-bridge/src/bridge.rs

```diff
@@ -0,0 +1,580 @@
+            for result in results {
+                for h in result? {
+                    headers.push(h);
+                }
+            }
+        }
+        Ok(headers)
+    }
+
+    fn get_header_future(block_range: Range<u64>) -> JoinHandle<anyhow::Result<Vec<FullHeader>>> {
+        tokio::spawn(async move {
+            debug!("Fetching headers: {block_range:?}");
+            let batch_request: Vec<JsonRequest> = block_range
+                .into_iter()
+                .map(|i| {
+                    // Geth requires block numbers to be formatted using the following padding.
+                    let block_param = format!("0x{i:01X}");
+                    let params = Params::Array(vec![json!(block_param), json!(true)]);
+                    json_request("eth_getBlockByNumber".to_string(), params, i as u32)
+                })
+                .collect();
+            let response = geth_batch_request(batch_request).await?;
+            let batch: FullHeaderBatch = serde_json::from_str(&response)?;
+            Ok(batch.headers)
+        })
+    }
+}
+
+/// Datatype to help track & report stats for bridge
+/// A single GossipBatch should never cross epoch boundaries.
+struct GossipBatch {
+    range: Range<u64>,
+    epoch_index: u64,
+    full_headers: Vec<FullHeader>,
+    hwp_count: u64,
+    bodies_count: u64,
+    receipts_count: u64,
+    start_time: time::Instant,
+}
+
+impl GossipBatch {
+    fn new(range: Range<u64>, epoch_index: u64, full_headers: Vec<FullHeader>) -> Self {
+        Self {
+            range,
+            epoch_index,
+            full_headers,
+            hwp_count: 0,
+            bodies_count: 0,
+            receipts_count: 0,
+            start_time: time::Instant::now(),
+        }
+    }
+
+    fn display_stats(&self) {
+        info!(
+            "Header Group: Range {:?} - HWP: {:?} - Bodies: {:?} - Receipts: {:?}",
+            self.range, self.hwp_count, self.bodies_count, self.receipts_count
+        );
+        info!("Group took: {:?}", time::Instant::now() - self.start_time);
+    }
+}
+
+async fn get_latest_block_number() -> anyhow::Result<u64> {
+    let params = Params::Array(vec![json!("latest"), json!(false)]);
+    let method = "eth_getBlockByNumber".to_string();
+    let request = json_request(method, params, 1);
+    let response = geth_batch_request(vec![request]).await?;
+    let response: Vec<Value> = serde_json::from_str(&response)?;
+    let result = response[0]
+        .get("result")
+        .ok_or_else(|| anyhow!("Unable to fetch latest block"))?;
+    let header: Header = serde_json::from_value(result.clone())?;
+    Ok(header.number)
+}
+
+/// Used the "surf" library here instead of "ureq" since "surf" is much more capable of handling
+/// multiple async requests. Using "ureq" consistently resulted in errors as soon as the number of
+/// concurrent tasks increased significantly.
+async fn geth_batch_request(obj: Vec<JsonRequest>) -> anyhow::Result<String> {
```

Comment:
Why is this called `geth`? Isn't it agnostic to the client, and calling the swarm of available clients in pandaops?